include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -597,36 +597,59 @@ template <typename number>
 void
 AffineConstraints<number>::make_consistent_in_parallel(
   const IndexSet &locally_owned_dofs,
-  const IndexSet &locally_relevant_dofs,
+  IndexSet        constraints_to_make_consistent,
   const MPI_Comm  mpi_communicator)
 {
   if (Utilities::MPI::n_mpi_processes(mpi_communicator) == 1)
     return; // nothing to do, since serial
 
   Assert(sorted == false, ExcMatrixIsClosed());
 
-  // 1) get all locally relevant constraints ("temporal constraint matrix")
-  const auto temporal_constraint_matrix =
-    internal::compute_locally_relevant_constraints(*this,
-                                                   locally_owned_dofs,
-                                                   locally_relevant_dofs,
-                                                   mpi_communicator);
+  IndexSet locally_stored_constraints;
+  IndexSet constraints_made_consistent;
+  do
+    {
+      // 1) get all locally relevant constraints ("temporal constraint matrix")
+      const auto temporal_constraint_matrix =
+        internal::compute_locally_relevant_constraints(
+          *this,
+          locally_owned_dofs,
+          constraints_to_make_consistent,
+          mpi_communicator);
+
+      // 2) add untracked dofs to the index sets
+      std::vector<types::global_dof_index> constrained_indices;
```

Comment:
Supposedly this can get pretty big for every iteration. Would it make sense to move the variable declaration outside the loop and only `clear()` it here?
Also, it might make sense to `reserve` enough space at least of the size of the `temporal_constraint_matrix` to avoid frequent reallocations in the loop below for the first iteration.